### PR TITLE
partner: Add Upstage partner package LA and GC

### PR DESCRIPTION
**Pull request by junkeon0412**
**Description:** add Upstage partner packages including UpstageLayoutAnalysis and UpstageGroundnessCheck
**Dependencies:** langchain_community
**Twitter handle:** [@upstageai](https://twitter.com/upstageai)

- [x] **Add tests and docs**: If you're adding a new integration, please include
  1. a test for the integration, preferably unit tests that do not rely on network access,
  2. an example notebook showing its use. It lives in `docs/docs/integrations` directory.

- [x] **Lint and test**: Run `make format`, `make lint` and `make test` from the root of the package(s) you've modified. See contribution guidelines for more: https://python.langchain.com/docs/contributing/

Additional guidelines:
- Make sure optional dependencies are imported within a function.
- Please do not add dependencies to pyproject.toml files (even optional ones) unless they are required for unit tests.
- Most PRs should not touch more than one package.
- Changes should be backwards compatible.
- If you are adding something to community, do not re-import it in langchain.

If no one reviews your PR within a few days, please @-mention one of baskaryan, efriis, eyurtsev, hwchase17.

### CODE REVIEW

File: .github/workflows/_integration_test.yml [hunk 1]
```diff
@@ -43,11 +43,11 @@ jobs:
...
         run: poetry run pip install "boto3<2" "google-cloud-aiplatform<2"
 
-      - name: 'Authenticate to Google Cloud'
-        id: 'auth'
+      - name: "Authenticate to Google Cloud"
+        id: "auth"
         uses: google-github-actions/auth@v2
         with:
-          credentials_json: '${{ secrets.GOOGLE_CREDENTIALS }}'
+          credentials_json: "${{ secrets.GOOGLE_CREDENTIALS }}"
```
Comment (chosh0615) [inline .github/workflows/_integration_test.yml:50]: Unnecessary changes?

File: .github/workflows/_release.yml [hunk 2]
```diff
@@ -12,7 +12,7 @@ on:
       working-directory:
         required: true
         type: string
-        default: 'libs/langchain'
+        default: "libs/langchain"
```
Comment (chosh0615) [inline .github/workflows/_release.yml:15]: Unnecessary change?

File: libs/partners/upstage/langchain_upstage/tools/groundedness_check.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+)
+from langchain_core.messages import AIMessage, BaseMessage, HumanMessage
+from langchain_core.pydantic_v1 import BaseModel, Field, SecretStr
+from langchain_core.tools import BaseTool
+
+from langchain_upstage import ChatUpstage
+
+
+class GroundednessCheckInput(BaseModel):
+    """Input for the Groundness Check tool."""
```
Comment (chosh0615) [inline libs/partners/upstage/langchain_upstage/tools/groundedness_check.py:16]:
```suggestion
    """Input for the Groundedness Check tool."""
```

File: libs/partners/upstage/langchain_upstage/tools/groundedness_check.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    context: str = Field(description="context in which the answer should be verified")
+    assistant_message: str = Field(
+        description="assistant's reply or any text that is subject to groundness check"
```
Comment (chosh0615) [inline libs/partners/upstage/langchain_upstage/tools/groundedness_check.py:20]:
```suggestion
        description="assistant's reply or a text that is subject to groundedness check"
```

File: libs/partners/upstage/tests/integration_tests/test_groundness_check.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+from langchain_upstage import GroundednessCheck
+
+
+def test_langchain_upstage_groundness_check() -> None:
```
Comment (chosh0615) [inline libs/partners/upstage/tests/integration_tests/test_groundness_check.py:4]:
```suggestion
def test_langchain_upstage_groundedness_check() -> None:
```

File: libs/partners/upstage/tests/integration_tests/test_groundness_check.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    """Test Upstage Groundedness Check."""
+    tool = GroundednessCheck()
+    output = tool.run({"context": "foo bar", "assistant_message": "bar foo"})
+
+    assert output.response_metadata["model_name"] == tool.api_wrapper.model_name
+
+
+async def test_langchain_upstage_groundness_check_async() -> None:
```
Comment (chosh0615) [inline libs/partners/upstage/tests/integration_tests/test_groundness_check.py:12]:
```suggestion
async def test_langchain_upstage_groundedness_check_async() -> None:
```

File: libs/partners/upstage/tests/integration_tests/test_groundness_check.py
Comment (chosh0615) [inline libs/partners/upstage/tests/integration_tests/test_groundness_check.py:1]: `groundedness_check` in the file name.

File: libs/partners/upstage/pyproject.toml [hunk 7]
```diff
@@ -12,24 +12,30 @@ license = "MIT"
...
+langchain_community = ">=0.0.31"
 
 [tool.poetry.group.test]
 optional = true
 
 [tool.poetry.group.test.dependencies]
 pytest = "^7.3.0"
 freezegun = "^1.2.2"
-pytest-mock = "^3.10.0"
+pytest-mock  = "^3.10.0"
```
Comment (chosh0615) [inline libs/partners/upstage/pyproject.toml:29]: revert

File: libs/partners/upstage/pyproject.toml [hunk 8]
```diff
@@ -12,24 +12,30 @@ license = "MIT"
...
 syrupy = "^4.0.2"
 pytest-watcher = "^0.3.4"
 pytest-asyncio = "^0.21.1"
-langchain-openai = { path = "../openai", develop = true }
-langchain-core = { path = "../../core", develop = true }
+langchain-openai = {path = "../openai", develop = true}
+langchain-core = {path = "../../core", develop = true}
```
Comment (chosh0615) [inline libs/partners/upstage/pyproject.toml:34]: revert

File: libs/partners/upstage/pyproject.toml [hunk 9]
```diff
@@ -50,26 +56,28 @@ ruff = "^0.1.5"
 
 [tool.poetry.group.typing.dependencies]
 mypy = "^0.991"
-langchain-core = { path = "../../core", develop = true }
+langchain-core = {path = "../../core", develop = true}
```
Comment (chosh0615) [inline libs/partners/upstage/pyproject.toml:59]: revert

File: libs/partners/upstage/pyproject.toml [hunk 10]
```diff
@@ -50,26 +56,28 @@ ruff = "^0.1.5"
...
 
 [tool.poetry.group.dev]
 optional = true
 
 [tool.poetry.group.dev.dependencies]
-langchain-core = { path = "../../core", develop = true }
+langchain-core = {path = "../../core", develop = true}
```
Comment (chosh0615) [inline libs/partners/upstage/pyproject.toml:65]: revert

File: libs/partners/upstage/pyproject.toml [hunk 11]
```diff
@@ -50,26 +56,28 @@ ruff = "^0.1.5"
...
 
 [tool.ruff]
 select = [
-  "E", # pycodestyle
-  "F", # pyflakes
-  "I", # isort
+  "E",  # pycodestyle
+  "F",  # pyflakes
+  "I",  # isort
```
Comment (chosh0615) [inline libs/partners/upstage/pyproject.toml:71]: revert

File: libs/partners/upstage/tests/integration_tests/test_embeddings.py [hunk 12]
```diff
@@ -1,4 +1,5 @@
 """Test Upstage embeddings."""
+
```
Comment (chosh0615) [inline libs/partners/upstage/tests/integration_tests/test_embeddings.py:2]: revert

File: libs/partners/upstage/tests/unit_tests/test_embeddings.py [hunk 13]
```diff
@@ -1,4 +1,5 @@
 """Test embedding model integration."""
+
```
Comment (chosh0615) [inline libs/partners/upstage/tests/unit_tests/test_embeddings.py:2]: revert

File: libs/partners/upstage/tests/unit_tests/test_chat_models_standard.py [hunk 14]
```diff
@@ -1,4 +1,5 @@
 """Standard LangChain interface tests"""
+
```
Comment (chosh0615) [inline libs/partners/upstage/tests/unit_tests/test_chat_models_standard.py:2]: revert

File: docs/docs/integrations/providers/upstage.ipynb [hunk 15]
```diff
@@ -35,7 +37,9 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Get an [access token](https://console.upstage.ai) and set it as an environment variable (`UPSTAGE_API_KEY`)"
+    "Get an [access token](https://console.upstage.ai) and set environment variables `UPSTAGE_API_KEY` and `UPSTAGE_DOCUMENT_AI_API_KEY`.\n",
```
Comment (chosh0615) [inline docs/docs/integrations/providers/upstage.ipynb:40]:
```suggestion
    "Get [API Keys](https://console.upstage.ai) and set environment variables `UPSTAGE_API_KEY` and `UPSTAGE_DOCUMENT_AI_API_KEY`.\n",
```

File: docs/docs/integrations/providers/upstage.ipynb [hunk 16]
```diff
@@ -35,7 +37,9 @@
...
+    "\n",
+    "> As of April 2024, you need separate access tokens for Solar and Document AI(Layout Analysis). The access tokens will be consolidated soon (hopefully in May) and you'll need just one key for all features."
```
Comment (chosh0615) [inline docs/docs/integrations/providers/upstage.ipynb:42]:
```suggestion
    "> As of April 2024, you need separate API Keys for Solar and Document AI(Layout Analysis). The API Keys will be consolidated soon (hopefully in May) and you'll need just one key for all features."
```

File: docs/docs/integrations/providers/upstage.ipynb [hunk 17]
```diff
@@ -120,6 +127,130 @@
...
+    "\n",
+    "from langchain_community.vectorstores import DocArrayInMemorySearch\n",
+    "from langchain_core.documents.base import Document\n",
+    "from langchain_core.output_parsers import StrOutputParser\n",
+    "from langchain_core.prompts import ChatPromptTemplate\n",
+    "from langchain_core.runnables import RunnablePassthrough\n",
+    "from langchain_core.runnables.base import RunnableSerializable\n",
+    "from langchain_upstage import ChatUpstage, LayoutAnalysis\n",
+    "from langchain_upstage.embeddings import UpstageEmbeddings\n",
+    "from langchain_upstage.tools.groundedness_check import GroundednessCheck\n",
```
Comment (chosh0615) [inline docs/docs/integrations/providers/upstage.ipynb:210]:
```suggestion
    "from langchain_upstage import ChatUpstage, GroundednessCheck, LayoutAnalysis, UpstageEmbeddings\n",
```

Wrong imports?

File: docs/docs/integrations/providers/upstage.ipynb [hunk 18]
```diff
@@ -120,6 +127,130 @@
...
+    "\n",
+    "model = ChatUpstage()\n",
+    "\n",
+    "file_path = \"/PATH/TO/YOUR/FILE.pdf\"\n",
```
Comment (chosh0615) [inline docs/docs/integrations/providers/upstage.ipynb:214]: Can we add one line that downloads a pdf file that contains Solar information here?

File: libs/partners/upstage/langchain_upstage/layout_analysis.py [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+        return default
+    else:
+        raise ValueError(
+            f"Did not find {key}, please add an environment variable"
+            f" `{env_key}` which contains it, or pass"
+            f"  `{key}` as a named parameter."
+        )
+
+
+class LayoutAnalysis:
```
Comment (chosh0615) [inline libs/partners/upstage/langchain_upstage/layout_analysis.py:68]: Do we need any base class?

File: libs/partners/upstage/tests/integration_tests/test_embeddings.py (hunk 12 above)
Comment (junkeon) [reply libs/partners/upstage/tests/integration_tests/test_embeddings.py:2]: it is because of linting

File: libs/partners/upstage/tests/unit_tests/test_embeddings.py (hunk 13 above)
Comment (junkeon) [reply libs/partners/upstage/tests/unit_tests/test_embeddings.py:2]: it is because of linting

File: libs/partners/upstage/tests/unit_tests/test_chat_models_standard.py (hunk 14 above)
Comment (junkeon) [reply libs/partners/upstage/tests/unit_tests/test_chat_models_standard.py:2]: it is because of linting

File: .github/workflows/_integration_test.yml [hunk 20]
```diff
@@ -79,6 +79,7 @@ jobs:
           VOYAGE_API_KEY: ${{ secrets.VOYAGE_API_KEY }}
           COHERE_API_KEY: ${{ secrets.COHERE_API_KEY }}
           UPSTAGE_API_KEY: ${{ secrets.UPSTAGE_API_KEY }}
+          UPSTAGE_DOCUMENT_AI_API_KEY: ${{ secrets.UPSTAGE_DOCUMENT_AI_API_KEY }}
```
Comment (efriis) [inline .github/workflows/_integration_test.yml:82]: is this needed? I don't think there are any integration tests for document ai

File: .github/workflows/_release.yml [hunk 21]
```diff
@@ -216,6 +216,7 @@ jobs:
           MONGODB_ATLAS_URI: ${{ secrets.MONGODB_ATLAS_URI }}
           VOYAGE_API_KEY: ${{ secrets.VOYAGE_API_KEY }}
           UPSTAGE_API_KEY: ${{ secrets.UPSTAGE_API_KEY }}
+          UPSTAGE_DOCUMENT_AI_API_KEY: ${{ secrets.UPSTAGE_DOCUMENT_AI_API_KEY }}
```
Comment (efriis) [inline .github/workflows/_release.yml:219]: is this needed? I don't think there are any integration tests for document ai

File: docs/docs/integrations/providers/upstage.ipynb
Comment (efriis) [inline docs/docs/integrations/providers/upstage.ipynb:1]: let's remove the RAG section and move that to a cookbook instead (it's a bit too long for the provider page)

**Review by efriis — COMMENTED** [review]
some comments. and the documentation build still might need some work.

File: docs/docs/integrations/providers/upstage.ipynb
Comment (chosh0615) [reply docs/docs/integrations/providers/upstage.ipynb:1]: Moved it to the cookbook!

File: .github/workflows/_integration_test.yml (hunk 20 above)
Comment (chosh0615) [reply .github/workflows/_integration_test.yml]: Removed!

File: .github/workflows/_release.yml (hunk 21 above)
Comment (chosh0615) [reply .github/workflows/_release.yml]: Removed!
